### PR TITLE
feat(bindings): add Java JNI bindings

### PR DESCRIPTION
**Pull request by freakynit**
## Summary

- add Java 21 bindings over the stable C API through a JNI bridge
- cover database and transaction lifecycle, graph CRUD/traversal, properties and indexes, Cypher queries, vector/full-text search, durable streams, query-cache diagnostics, and embedding helpers
- add a Maven build that compiles and stages the native libraries on Linux and macOS
- document the Java API and include a runnable knowledge-graph example
- run the Java suite in CI on Ubuntu and macOS

## Implementation disclosure

The Java binding was produced with AI assistance. Before opening this PR, the generated implementation was reviewed and corrected for clean-build reproducibility, Linux/macOS portability, JNI reference and allocation safety, API edge cases, documentation accuracy, and example execution.

## Testing

- Maven clean test: 33 tests passed, covering lifecycle/read-only behavior and rollback, node/edge operations, nested values and large collections, property indexes, query parameters and diagnostics, vector and full-text search, streams/changefeeds, and query-cache behavior
- repository Zig suite: 749 tests passed and 5 skipped (7/7 build steps succeeded)
- JNI bridge compiled with additional warning checks
- documented knowledge-graph example executed successfully against the staged native library

The Java CI matrix exercises clean Maven builds on both Ubuntu and macOS so the native staging and platform-specific loader configuration stay covered.

### CODE REVIEW

**Comment by jeffhajewski** [comment]
Thanks for the PR! A few things to fix before we can merge.

- Rebase on the latest release (`0.13.0`). Also of note, the new release introduces a file lock mechanism, which throws a new error code, `-16`. This will need to be incorporated into your error handling.
- `ErrorCode.fromCode` should fall back to a catch all that preserves the numeric code rather than throwing. We want to provide as much diagnostic info to the user as possible.
- `Database.close()` is synchronized but `handle()` is not, so one thread can read a live handle, another can close and free it, and the first then uses a stale `jlong`. The C API's handle registry catches this and returns `err_invalid_arg` rather than dereferencing freed memory — so it's not the use-after-free it first looks like — but address reuse could still route an operation to a different database. Making `handle()` synchronized removes the class of problem.

**Comment by freakynit** [comment]
Thanks for the review! All three points are addressed in 5861ba4:

- Rebased on latest main (0.13.0 onward). The new `LATTICE_ERROR_DATABASE_LOCKED` (`-16`) is now mapped to a new `ErrorCode.FILE_LOCKED` enum constant, distinct from `LOCK_TIMEOUT`. The JNI bridge still compiles cleanly against the new headers (`lattice_open_v3` remains available), so no ABI changes were needed.
- `ErrorCode.fromCode` no longer throws — it falls back to an `UNKNOWN` catch-all, and `LatticeException` now carries the raw numeric code via `getNativeCode()` (and includes it in the message) so diagnostics are preserved.
- `Database.handle()` is now `synchronized`, closing the close-vs-handle race.

All 35 tests pass locally (including two new ones covering the catch-all fallback and the -16 mapping).

File: bindings/java/src/main/java/io/latticedb/OpenOptions.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
+package io.latticedb;
+
+import java.util.Objects;
+
+/**
+ * Options for opening a database, mirroring the Go binding's OpenOptions and
+ * the native {@code lattice_open_options_v3} struct. All values have sane
+ * defaults; use the builder to override them.
+ */
+public final class OpenOptions {
```
Comment (jeffhajewski) [inline bindings/java/src/main/java/io/latticedb/OpenOptions.java:10]: we should expose `lock` like the other bindings

File: bindings/java/src/main/java/io/latticedb/ErrorCode.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    FULL(-10),
+    VERSION_MISMATCH(-11),
+    CHECKSUM(-12),
+    OUT_OF_MEMORY(-13),
+    UNSUPPORTED(-14),
+    VALUE_TOO_LARGE(-15),
+    /** Another process holds the database file (native code -16, added in
+     * 0.13.0). Distinct from {@link #LOCK_TIMEOUT}, which reports a second
+     * writer inside this process. */
+    FILE_LOCKED(-16),
```
Comment (jeffhajewski) [inline bindings/java/src/main/java/io/latticedb/ErrorCode.java]: nit: `DATABASE_LOCKED`

**Comment by jeffhajewski** [comment]
A few more comments

- In v0.14.0 we added `serialize/deserialize` - these need to be added
- `FindClass` is uncached — 20 calls in the value-conversion path, inconsistent with their own `CACHE_BOX` caching in `JNI_OnLoad`. This is a perf issue, not correctness

File: bindings/java/pom.xml [hunk 3]
```diff
@@ -0,0 +1,137 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>io.latticedb</groupId>
+  <artifactId>latticedb</artifactId>
+  <version>0.12.0</version>
```
Comment (jeffhajewski) [inline bindings/java/pom.xml]: nit: right now this is `0.14.0`. By EOD there's a good chance it will be `0.15.0`

**Comment by freakynit** [comment]
@jeffhajewski Addressed all follow-up feedback in c0f8b82:

- Rebased onto current origin/main
- Exposed OpenOptions.lock(boolean) using lattice_open_options_v4
- Renamed FILE_LOCKED to DATABASE_LOCKED
- Added Database.serialize() / Database.deserialize(...)
- Cached JNI classes and method IDs during JNI_OnLoad
- Updated the Java artifact version to the current repository version, 0.14.0
- Added coverage for snapshot round-tripping, invalid serialized data, transaction locking, and option defaults

mvn clean test passes all 38 tests locally.

**Comment by freakynit** [comment]
@jeffhajewski Sorry to ping here again, but since the code is now merged, I have worked on updating relevant documentation (`book`). 

Here are the changes: https://github.com/jeffhajewski/latticedb/compare/main...freakynit:latticedb:docs/add-java-api-reference?expand=1

If it seems okay, I can raise a new PR for this.

I have intentionally skipped updating the main page since java library is not yet published.

Thanks..

**Comment by jeffhajewski** [comment]
> @jeffhajewski Sorry to ping here again, but since the code is now merged, I have worked on updating relevant documentation (`book`).
> 
> Here are the changes: https://github.com/jeffhajewski/latticedb/compare/main...freakynit:latticedb:docs/add-java-api-reference?expand=1
> 
> If it seems okay, I can raise a new PR for this.
> 
> I have intentionally skipped updating the main page since java library is not yet published.
> 
> Thanks..

Of course! You never have to ask permission to raise a PR.
